feat: nodes scan wifi channels during discovery

jamienz opened the pull request:
When a node is performing the discovery process, it loops through an array of wifi channels and sends its broadcast on each channel.  Its possible that the node sends a broadcast on one channel, and the host receives it despite being on another (due to the fact that wifi channels overlap?), so the host includes its channel in the response.  The node then saves this channel in NVS.

jamienz committed:
feat: nodes scan wifi channels during discovery

Johboh commented:
Thanks for you contribution @jamienz! I didn't expect anyone to use this over engineered library other than my self, given the sparse documentation :).

I'm going to cherry pick this and try it out, and I will leave some inline comments as well.

Do you know, how does the WiFi channel relate to the ESP-NOW channel? As we are setting [here](https://github.com/Johboh/EspNowNetwork/blob/main/src/impl/EspNowHost.cpp#L256) and [here](https://github.com/Johboh/EspNowNetwork/blob/main/src/impl/EspNowNode.cpp#L124). E.g. if we should set this instead/also.

For my "statistics", do you use the ESP-IDF or Arduino library variant?

jamienz commented:
Thanks for creating the EspNowNetwork library @Johboh!

When you do your testing of this PR, make sure that the [WiFi network](https://github.com/Johboh/EspNowNetwork/blob/e4f5efaa4177ada93932981f90c07cea74314b37/examples/arduino/host/Host.ino#L22) that the host is connected to is NOT on channel 1.

When the host is connected to a WiFi network, it will send/receive ESP-NOW messages on the same channel as that network (it can't send on a different channel).  The node isn't connected to a WiFi network, so by default it will be on channel 1, based on my testing by adding the following code to the node before it sends it's discovery broadcasts:
```
uint8_t primary = 0;
wifi_second_chan_t second;
ESP_ERROR_CHECK(esp_wifi_get_channel(&primary, &second));
log("channel=" + std::to_string(primary), ESP_LOG_INFO);
```

The peer_info.channel is set to zero which means ESP-NOW messages will be sent on whatever channel is the current WiFi channel - for the host it will be the WiFi network channel, and for the node it will be 1.  With the code currently in the main branch, everything will work well if the host is connected to a WiFi network on channel 1.  this PR aims to support the host being on a channel other than channel 1.

I'm using PlatformIO/Arduino.

Johboh commented:
Right, I was just thinking of what the point was with the [channel](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_now.html#_CPPv4N17esp_now_peer_info7channelE) field in the peer info. If I understand the documentation correctly, if set to 0 it will use same channel "which station or softap is on", otherwise if not zero it must be set to be "the channel that station or softap is on.". So if I'm not missing something, the [channel](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_now.html#_CPPv4N17esp_now_peer_info7channelE) is quite useless and could have been omitted in the ESP-now API?

Johboh commented on src/EspNowPreferences.h:
```diff
@@ -34,6 +34,19 @@ class EspNowPreferences : public EspNowNetwork::Preferences {
+  bool espNowSetChannelForHost(uint8_t channel) override;
+
+  /**
+   * @brief Return the WiFi channel stored, or std::nullopt if no channel stored.
```
```suggestion
   * @brief Get the WiFi channel stored. Returns false if no channel stored.
```
(I know you copy-pasted this from above, which is also wrong. I copy-pasted this code form another project where I use std::optional as an return argument instead of having a bool return and pointer as arg)

jamienz replied:
fixed

Johboh commented on src/Preferences.h:
```diff
@@ -26,6 +26,19 @@ class Preferences {
+  virtual bool espNowSetChannelForHost(uint8_t channel) = 0;
+  
+  /**
+   * @brief Return the WiFi channel stored, or std::nullopt if no channel stored.
```
```suggestion
   * @brief Get the WiFi channel stored. Returns false if no channel stored.
```

jamienz replied:
fixed

Johboh commented on src/impl/EspNowNode.cpp:
```diff
@@ -23,6 +22,9 @@
 #define NUMBER_OF_RETRIES_FOR_CHALLENGE_REQUEST 50
 
+// wifi channels that will be tried during discovery
+#define WIFI_SCAN_CHANNELS {1,2,3,4,5,6,7,8,9,10,11,12,13,14}
```
This doesn't look like it have been formatted with clang-format. Could it be so that you don't have clang-format enabled/installed?

jamienz replied:
You're right I don't have that installed. I'll install it.

jamienz replied:
fixed

Johboh commented on src/impl/EspNowNode.cpp:
```diff
@@ -84,8 +87,9 @@ bool EspNowNode::setup() {
   ESP_ERROR_CHECK(esp_wifi_init(&cfg));
   ESP_ERROR_CHECK(esp_wifi_set_storage(WIFI_STORAGE_RAM));
   ESP_ERROR_CHECK(esp_wifi_set_mode(WIFI_MODE_STA));
+  ESP_ERROR_CHECK(esp_wifi_set_ps(WIFI_PS_NONE));
```
Why is this one needed?

jamienz replied:
That makes sure wifi doesn't go into power saving mode, which can result in it not receiving ESP-NOW frames. However, given the node wakes up gor short period of time, there probably isn't time for power saving to kick in on the node, so this could be removed. It should stay on the host though.

jamienz replied:
I've removed it given my comment above.

Johboh commented on src/impl/EspNowNode.cpp:
```diff
@@ -131,6 +135,12 @@ bool EspNowNode::setup() {
+    uint8_t channel = 1;
+    _preferences.espNowGetChannelForHost(&channel);
+    log("loaded channel " + std::to_string(channel), ESP_LOG_INFO);
+    ESP_ERROR_CHECK(esp_wifi_set_promiscuous(true));
```
Suggest moving these three repeated lines (used here and L:180) into a class member function (set_promiscuous, set_channel, set_promiscuous)

jamienz replied:
I tested changing the channel without enabling/disabling promiscuous mode, and it worked.  That enable/disable was something that was necessary in the past, although that may have been when the device is also connected to a WiFi network.  But either way, testing without enable/disable was fine, so I've removed those lines, and that removes the need for the separate function.

Johboh commented on src/impl/EspNowNode.cpp:
```diff
@@ -159,7 +169,18 @@ bool EspNowNode::setup() {
 
     int8_t retries = NUMBER_OF_RETRIES_FOR_DISCOVERY_REQUEST;
+
+    uint8_t channels[] = WIFI_SCAN_CHANNELS;
```
Instead of assigning, you could declare `uint8_t WIFI_SCAN_CHANNELS[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14};` at the top instead of the existing `WIFI_SCAN_CHANNELS` define.

Johboh replied:
@jamienz And I suggest doing this as well.

Johboh commented on src/impl/EspNowNode.cpp:
```diff
@@ -159,7 +169,18 @@ bool EspNowNode::setup() {
     int8_t retries = NUMBER_OF_RETRIES_FOR_DISCOVERY_REQUEST;
+
+    uint8_t channels[] = WIFI_SCAN_CHANNELS;
+    uint8_t channelCount = sizeof(channels);
```
Use snake_case for variable names. (`channel_count`)

Johboh replied:
@jamienz this still apply.

Johboh commented on src/impl/EspNowNode.cpp:
```diff
@@ -159,7 +169,18 @@ bool EspNowNode::setup() {
+
     while (retries-- > 0) {
+      uint8_t channel = channels[channelIdx++ % channelCount];
+      log("setting discovery channel " + std::to_string(channel), ESP_LOG_INFO);
```
To reduce log rows, you could include this in the `log("Sending broadcast discovery request on channel...` just below

Johboh replied:
@jamienz And I suggest doing this as well.

Johboh commented on src/impl/EspNowPreferences.cpp:
```diff
@@ -9,6 +9,7 @@
 #define NVS_STORAGE "storage"
 // Max key length: 15 chars
 #define NVS_STORAGE_KEY_HOST_MAC "host_mac"
+#define NVS_STORAGE_KEY_HOST_CHAN "host_chan"
```
```suggestion
#define NVS_STORAGE_KEY_HOST_CHAN "host_channel"
```

jamienz replied:
fixed

Johboh commented on src/impl/EspNowPreferences.cpp:
```diff
@@ -30,6 +31,43 @@ void EspNowPreferences::initalizeNVS() {
 
+bool EspNowPreferences::espNowSetChannelForHost(uint8_t channel) {
+  auto key = NVS_STORAGE_KEY_HOST_CHAN;
+  esp_err_t err = nvs_set_blob(_nvs_handle, key, &channel, 1);
```
Use `nvs_set_u8` instead.

jamienz replied:
fixed

Johboh commented on src/impl/EspNowPreferences.cpp:
```diff
@@ -30,6 +31,43 @@ void EspNowPreferences::initalizeNVS() {
+    return false;
+  }
+
+  size_t required_size;
```
Use `nvs_get_u8` instead. Then you don't need to reserve any size.
Example:
```
  auto key = NVS_STORAGE_KEY_HOST_CHAN;
  esp_err_t err = nvs_get_u8(_nvs_handle, key.c_str(), &buffer);
  if (err != ESP_OK) {
    ESP_LOGE(TAG, "Failed to get u8 from NVS with key %s (%s)", key.c_str(), esp_err_to_name(err));
    return false;
  }
  return true;
```

jamienz replied:
fixed

Johboh commented on src/EspNowPreferences.h:
```diff
@@ -34,6 +34,19 @@ class EspNowPreferences : public EspNowNetwork::Preferences {
+   *
+   * @param return true if channel was read successfully.
+   */
+  bool espNowGetChannelForHost(uint8_t *buffer) override;
```
```suggestion
  bool espNowGetChannelForHost(uint8_t *channel) override;
```

jamienz replied:
fixed

Johboh commented on src/Preferences.h:
```diff
@@ -26,6 +26,19 @@ class Preferences {
+   *
+   * @param return true if channel was read successfully.
+   */
+  virtual bool espNowGetChannelForHost(uint8_t *buffer) = 0;
```
```suggestion
  virtual bool espNowGetChannelForHost(uint8_t *channel) = 0;
```

jamienz replied:
fixed

Johboh commented:
Added some comments! Sorry for being nit-picky, but I review PRs for a living :)

Johboh commented:
(I moved some files around, but no change to the files themself, so a git pull --rebase should resolve the merge issue automatically)

jamienz commented:
Thanks for the feedback. I'll make those changes tonight.

jamienz committed:
Merge branch 'main' into feature/channel-scanning

jamienz committed:
fix: review fixes

jamienz committed:
fix: feedback fixes

jamienz committed:
fix: simplified setting wifi channel

enabling/disabling promiscuous mode doesn't appear to be needed any
more, having tested without it

jamienz committed:
fix: comment fix

Johboh commented:
Some minor comments left, and a failing clang-format (whitespace). If those are fixed, we are good to merge.

jamienz committed:
fix: clang format

jamienz commented:
> Right, I was just thinking of what the point was with the [channel](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_now.html#_CPPv4N17esp_now_peer_info7channelE) field in the peer info. If I understand the documentation correctly, if set to 0 it will use same channel "which station or softap is on", otherwise if not zero it must be set to be "the channel that station or softap is on.". So if I'm not missing something, the [channel](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_now.html#_CPPv4N17esp_now_peer_info7channelE) is quite useless and could have been omitted in the ESP-now API?

Thats a good question.  Its definitely not useless, its doing something, but I'll admit that I don't understand exactly what its doing - the documentation is incomplete in my opinion.  

If peer_info.channel is assigned zero, and the Node wifi channel isn't explicitly set, and the Node doesn't connect to a wifi network, the wifi channel seems to be 1.

My WiFi access point is on channel 6 right now, and since the Host connects to that access point, it is on channel 6 also.  So when the Node broadcasts on channel 1, the Host doesn't hear it.  Setting the peer_info.channel on the Node to 6 doesn't work:
`E (280) ESPNOW: Peer channel is not equal to the home channel, send fail!`
The Node needs to be on Wifi channel 6 also.

See: https://github.com/espressif/esp-idf/issues/9592.  The important final comment from the Espressif guy is "ESP-NOW's channel should obey the STA/AP's channel".  Which I interpret to mean:
* the Node's peer_info.channel needs to be 0, so that it uses the Node's WiFi channel (which needs to match the Host's WiFi channel)
* or the Node's peer_info.channel needs to be explicitly set to the WiFi channel

Leaving peer_info.channel=0 is the least-effort option.

Johboh commented:
@jamienz There are three more inline comments that it looks like you are missing:
https://github.com/Johboh/EspNowNetwork/pull/1#discussion_r1473023455
https://github.com/Johboh/EspNowNetwork/pull/1#discussion_r1473024817
https://github.com/Johboh/EspNowNetwork/pull/1#discussion_r1473029550

Johboh commented:
I will go a-head and merge this one as I need myself (I will address outstanding comments and fix clang format). There is also an unrecoverable crash loop that can happen in the node that I will fix (and that I ended up with when testing). Will share the fix commit here.

Johboh commented:
@jamienz FYI: https://github.com/Johboh/EspNowNetwork/commit/368aa095fd90f4ce5552674b872146782428128b

jamienz commented:
That's great @Johboh. I really appreciate your time in reviewing this, and for finishing it. Apologies for not getting back to this one sooner to finish off the final tasks.
